File: src/schemas/publish.ts
import { b } from "@zorsh/zorsh";

export class ImageExportOptions {
  frameSize?: [number, number];
  cellSize: number;
  preservedOverlap?: number;
  showGridLineNumbers: boolean;
  showCenteringMarks: boolean;

  constructor(data: b.infer<typeof ImageExportOptions.schema>) {
    if (data.frameSize) this.frameSize = data.frameSize;
    this.cellSize = data.cellSize;
    if (data.preservedOverlap) this.preservedOverlap = data.preservedOverlap;
    this.showGridLineNumbers = data.showGridLineNumbers;
    this.showCenteringMarks = data.showCenteringMarks;
  }

  static readonly schema = b.struct({
    frameSize: b.option(b.tuple(b.u16(), b.u16())),
    cellSize: b.f32(),
    preservedOverlap: b.option(b.u16()),
    showGridLineNumbers: b.bool(),
    showCenteringMarks: b.bool(),
  });

  static default(): ImageExportOptions {
    return new ImageExportOptions({
      frameSize: null,
      cellSize: 14.0,
      preservedOverlap: 3,
      showGridLineNumbers: false,
      showCenteringMarks: false,
    });
  }
}

export class PdfExportOptions {
  monochrome: boolean;
  color: boolean;
  centerFrames: boolean;
  enumerateFrames: boolean;
  frameOptions: ImageExportOptions;

  constructor(data: b.infer<typeof PdfExportOptions.schema>) {
    this.monochrome = data.monochrome;
    this.color = data.color;
    this.centerFrames = data.centerFrames;
    this.enumerateFrames = data.enumerateFrames;
    this.frameOptions = new ImageExportOptions(data.frameOptions);
  }

  static readonly schema = b.struct({
    monochrome: b.bool(),
    color: b.bool(),
    centerFrames: b.bool(),
    enumerateFrames: b.bool(),
    frameOptions: ImageExportOptions.schema,
  });

  static default(): PdfExportOptions {
    return new PdfExportOptions({
      monochrome: true,
      color: false,
      centerFrames: false,
      enumerateFrames: true,
      frameOptions: {
        frameSize: [30, 40],
        cellSize: 14.0,
        preservedOverlap: 3,
        showGridLineNumbers: true,
        showCenteringMarks: true,
      },
    });
  }
}

export class PublishSettings {
  pdf: PdfExportOptions;

  constructor(data: b.infer<typeof PublishSettings.schema>) {
    this.pdf = new PdfExportOptions(data.pdf);
  }

  static readonly schema = b.struct({
    pdf: PdfExportOptions.schema,
  });
}

File: app/src-tauri/tests/publish.rs
use embroiderly::commands;
use embroiderly::state::{HistoryState, PatternsState};
use embroiderly_pattern::{ImageExportOptions, PdfExportOptions};
use tauri::Manager as _;

mod utils;

#[test]
fn updates_pdf_export_options() {
  let (app, webview) = setup_test_app!(commands: [commands::core::publish::update_pdf_export_options]);
  let pattern_id = utils::create_test_pattern(&app);

  let new_pdf_options = PdfExportOptions {
    monochrome: false,
    color: true,
    center_frames: true,
    enumerate_frames: false,
    frame_options: ImageExportOptions {
      frame_size: Some((50, 60)),
      cell_size: 20.0,
      preserved_overlap: Some(5),
      show_grid_line_numbers: false,
      show_centering_marks: true,
    },
  };
  assert!(
    invoke_ipc!(
      &webview,
      cmd: "update_pdf_export_options",
      body: tauri::ipc::InvokeBody::Raw(borsh::to_vec(&new_pdf_options).unwrap()),
      headers: [("patternId", pattern_id.to_string().parse().unwrap())]
    )
    .is_ok()
  );

  let patterns_state = app.state::<PatternsState>();
  let patterns_manager = patterns_state.read().unwrap();

  let patproj = patterns_manager.get_pattern_by_id(&pattern_id).unwrap();
  assert_eq!(patproj.publish_settings.pdf, new_pdf_options);

  let history_state = app.state::<HistoryState<tauri::test::MockRuntime>>();
  let history_manager = history_state.read().unwrap();

  let history = history_manager.get(&pattern_id).unwrap();
  assert_eq!(history.undo_stack_len(), 1);
}
